File: 2919-minimum-increment-operations-to-make-array-beautiful/2919-minimum-increment-operations-to-make-array-beautiful.cpp
class Solution {
public:
    long long minIncrementOperations(vector<int>& nums, int k) {
        int n=nums.size();
        vector<long long> dp(n+3,0);
        for(int idx=n-1 ; idx>=0; idx--){
            long long ans=1e14;
            for(int d=0 ; d<3 ; d++){
                if(idx-d>=0){
                    ans=min(ans,max(0LL,k-(long long)nums[idx-d])+dp[idx+3-d]);
                }
            }
            dp[idx]=ans;
        }
        return min(dp[0],min(dp[1],dp[2]));
    }
};

File: 2919-minimum-increment-operations-to-make-array-beautiful/2919-minimum-increment-operations-to-make-array-beautiful.py
class Solution:
    def minIncrementOperations(self, nums: List[int], k: int) -> int:
        if len(nums)==3:
            if max(nums)>=k:
                return 0
            else:
                return k-max(nums)
        @cache
        def solve(index):
            if index+3>len(nums):
                return 0
            ans=float('inf')
            for i in range(index,min(len(nums),index+3)):
                if nums[i]<k:
                    ans=min(ans,k-nums[i]+solve(i+1))
                else:
                    ans=min(ans,solve(i+1))
            return ans
        x=solve(0)
        if x>=float('inf'):return 0
        return x